### PR TITLE
I18n/init db

### PR DESCRIPTION
**Pull request by alexandrebodin**
Init some migration & schema changes with i18n.

### CODE REVIEW

File: packages/strapi-database/lib/migration-manager.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  register(migration) {
+    debug('Register migration');
+    this.migrations.push(migration);
+  }
+
+  async runMigration(fn, options, context = {}) {
+    debug('Run migration');
+    await this.runBefore(options, context);
+    await await fn(options, context);
```
Comment (Convly) [inline packages/strapi-database/lib/migration-manager.js:20]: What's the point of this double `await` since `migrateSchema` doesn't return anything? Can't you just put a single await here?

File: packages/strapi-database/lib/migration-manager.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      if (isFunction(migration.before)) {
+        await migration.before(options, context);
+      }
+    }
+  }
+
+  async runAfter(options, context) {
+    debug('Run after migrations');
+
+    for (const migration of this.migrations.slice(0).reverse()) {
```
Comment (Convly) [inline packages/strapi-database/lib/migration-manager.js:37]: Why do you need to `slice(0)` here? What does it change from `this.migrations.reverse()`?

File: packages/strapi/lib/Strapi.js [hunk 3]
```diff
@@ -414,35 +416,48 @@ class Strapi {
...
+    const execFunction = async fn => {
       if (!fn) return;
 
       return fn();
     };
 
-    // plugins bootstrap
+    const configPath = `functions.${functionName}`;
+
+    // plugins register
```
Comment (Convly) [inline packages/strapi/lib/Strapi.js:428]: Here we have `plugins register`, `plugin bootstraps` and `pluginFunc`. Shouldn't all three be named `plugins functions`, `pluginFunctions`, and `pluginFunction` instead?

File: packages/strapi/lib/Strapi.js [hunk 4]
```diff
@@ -414,35 +416,48 @@ class Strapi {
...
         strapi.log.error(err);
         strapi.stop();
       });
     });
+
     await Promise.all(pluginBoostraps);
 
-    // user bootstrap
-    await execBootstrap(_.get(this.config, ['functions', 'bootstrap']));
+    // user register
```
Comment (Convly) [inline packages/strapi/lib/Strapi.js:441]: Here and `L.444`, did you meant `user` & `admin` **function** instead of `register`?

File: packages/strapi-utils/lib/content-types.js [hunk 5]
```diff
@@ -67,7 +67,12 @@ const getWritableAttributes = (model = {}) => {
 };
 
 const getNonVisibleAttributes = model => {
-  return _.uniq([model.primaryKey, ...getTimestamps(model), ...NON_VISIBLE_ATTRIBUTES]);
+  return _.uniq([
+    model.primaryKey,
+    ...NON_VISIBLE_ATTRIBUTES,
+    ...getTimestamps(model),
+    ...getNonWritableAttributes(model),
```
Comment (Convly) [inline packages/strapi-utils/lib/content-types.js:74]: So if an attribute is non-writable, then it shouldn't be visible too, right?

**Review by Convly — COMMENTED** [review]
Nice, just some questions :slightly_smiling_face:

File: packages/strapi-database/lib/migration-manager.js (hunk 1 above)
Comment (alexandrebodin) [reply packages/strapi-database/lib/migration-manager.js:20]: just a typo it makes not sense

File: packages/strapi-database/lib/migration-manager.js (hunk 2 above)
Comment (alexandrebodin) [reply packages/strapi-database/lib/migration-manager.js:37]: reverse mutates the object :)

File: packages/strapi-plugin-i18n/services/__tests__/localizations.test.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    title: {
+      type: 'string',
+      pluginOptions: {
+        i18n: {
+          localized: true,
+        },
+      },
+    },
+    stars: {
+      type: 'interger',
```
Comment (petersg83) [inline packages/strapi-plugin-i18n/services/__tests__/localizations.test.js:17]: integer*

File: packages/strapi-plugin-i18n/services/localizations.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    Object.assign(entry, { localizations });
+  }
+};
+
+const updateNonLocalizedFields = async (entry, { model }) => {
+  const fieldsToUpdate = pick(getNonLocalizedFields(model), entry);
+
+  if (Array.isArray(entry.localizations)) {
+    const updateQUeries = entry.localizations
```
Comment (petersg83) [inline packages/strapi-plugin-i18n/services/localizations.js:19]: updateQueries*

File: packages/strapi-database/lib/migration-manager.js (hunk 2 above)
Comment (Convly) [reply packages/strapi-database/lib/migration-manager.js:37]: Makes sense, didn't think about this :+1:

File: packages/strapi-utils/lib/content-types.js (hunk 5 above)
Comment (petersg83) [inline packages/strapi-utils/lib/content-types.js:74]: Why this change ?

File: packages/strapi-utils/lib/content-types.js [hunk 8]
```diff
@@ -173,9 +178,13 @@ const getGlobalId = (model, modelName, prefix) => {
   return model.globalId || _.upperFirst(_.camelCase(globalId));
 };
 
+const isRelationalAttribute = attribute =>
+  _.has(attribute, 'model') || _.has(attribute, 'collection');
+
```
Comment (petersg83) [inline packages/strapi-utils/lib/content-types.js:183]: Is an attribute either `scalar` or `relational`? If yes, was about using `return !isScalarAttribute()` ?

File: packages/strapi/lib/Strapi.js [hunk 9]
```diff
@@ -414,35 +416,48 @@ class Strapi {
...
     });
   }
 
+  async runBootstrapFunctions() {
+    return this.runLifecyclesFunctions('bootstrap');
+  }
+
+  async runRegisterFunctions() {
+    return this.runLifecyclesFunctions('register');
+  }
```
Comment (petersg83) [inline packages/strapi/lib/Strapi.js:459]: What is the purpose of splitting the bootstrap files in 2 (bootstrap & register)? Is it to run all the registrations after the bootstrap? If yes, what usecase do you have in mind?

File: packages/strapi-plugin-i18n/config/functions/register.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+      });
+
+      _.set(model.attributes, 'locale', {
+        writable: true,
+        private: false,
+        configurable: false,
+        type: 'string',
+      });
+    }
+  });
```
Comment (petersg83) [inline packages/strapi-plugin-i18n/config/functions/register.js:23]: I guess it is here for the moment as it is the fastest way. However do we agree it will move in a better place in V4?

**Review by petersg83 — CHANGES_REQUESTED** [review]

File: packages/strapi-plugin-i18n/config/functions/bootstrap.js [hunk 11]
```diff
@@ -14,4 +16,24 @@ const actions = ['create', 'read', 'update', 'delete'].map(uid => ({
...
+        }
+      });
+
+      _.set(model, 'lifecycles.afterCreate', async entry => {
+        await getService('localizations').addLocalizations(entry, { model });
+      });
+
+      _.set(model, 'lifecycles.afterUpdate', async entry => {
+        await getService('localizations').updateNonLocalizedFields(entry, { model });
+      });
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/strapi-plugin-i18n/config/functions/bootstrap.js:36]: I'm doing the whole lifecycles in another PR :)

[the remark this answers is not in the document]
Comment (petersg83) [reply packages/strapi-plugin-i18n/config/functions/bootstrap.js:36]: Perfect :)

File: packages/strapi-utils/lib/content-types.js (hunk 8 above)
Comment (alexandrebodin) [reply packages/strapi-utils/lib/content-types.js:183]: In case we add new types it woudl break. better to have eahc type with their own definition

File: packages/strapi/lib/Strapi.js (hunk 9 above)
Comment (alexandrebodin) [reply packages/strapi/lib/Strapi.js:459]: we need the register to change the content types before the db is initialized. you register then bootstrap. This will allow even more things in the futur

File: packages/strapi-plugin-i18n/config/functions/register.js (hunk 10 above)
Comment (alexandrebodin) [reply packages/strapi-plugin-i18n/config/functions/register.js:23]: I don't understand the question

File: packages/strapi-utils/lib/content-types.js (hunk 5 above)
Comment (alexandrebodin) [reply packages/strapi-utils/lib/content-types.js:74]: the const is actually equal to the NON writable const + published at. In fact we need to add all the non writable attributes. Until no they were static so the const was enough. but now we have locale et localizations that are dynamic so I added the non writable attributes dynamically

Comment (alexandrebodin) [reply packages/strapi-utils/lib/content-types.js:74]: Yep, more details answered lower

File: packages/strapi/lib/Strapi.js (hunk 9 above)
Comment (petersg83) [reply packages/strapi/lib/Strapi.js:459]: Should line 453 and 457 be inversed then ?
Happy to talk about it IRL

Comment (alexandrebodin) [reply packages/strapi/lib/Strapi.js:459]: Will remove those two functions actually

**Review by petersg83 — APPROVED** [review]
LGTM :)
